components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaDeviceConstellation.kt

```diff
@@ -43,31 +42,43 @@ class FxaDeviceConstellation(
 
     override fun state(): ConstellationState? = constellationState
 
-    override fun initDeviceAsync(
-        name: String,
-        type: DeviceType,
-        capabilities: Set<DeviceCapability>
-    ): Deferred<Boolean> {
-        return scope.async {
-            handleFxaExceptions(logger, "initializing device") {
-                account.initializeDevice(name, type.into(), capabilities.map { it.into() }.toSet())
-            }
-        }
+    private enum class DeviceFinalizeAction {
+        Initialize,
+        EnsureCapabilities,
+        None
     }
 
-    override fun ensureCapabilitiesAsync(capabilities: Set<DeviceCapability>): Deferred<Boolean> {
-        return scope.async {
-            handleFxaExceptions(logger, "ensuring capabilities") {
-                account.ensureCapabilities(capabilities.map { it.into() }.toSet())
+    override suspend fun finalizeDevice(authType: AuthType, config: DeviceConfig) = withContext(scope.coroutineContext) {
+        val finalizeAction = when (authType) {
+            AuthType.Signin,
+            AuthType.Signup,
+            AuthType.Pairing,
+            AuthType.MigratedCopy -> DeviceFinalizeAction.Initialize
```

Comment:
Hrm, the fact that there are so many ways to spell "did a fresh signin" seems like a tiny bit of a code-smell to me. What things do we need to do that differ between `Signin`, `Signup` and `Pairing`? (I seem to recall this being added initially for metrics purposes?)

Reply:
Another place where we use this is to kick-off sync after an `onAuthenticated` event (`first sync` vs `startup sync`), with a different mapping from device's init/ensure.

Otherwise, this is plumbed around in such detail mostly for telemetry, yeah.

Reply:
I'll think about perhaps a different internal/external representation of this.